### PR TITLE
[Infineon] SoftwareDiagnostic update

### PR DESCRIPTION
**Pull request by AlvinHsiao**
Problem
Infineon CYW30739 doesn't support DGSW.S.F00(Watermarks)

Change overview
Add DGSW.S.F00(Watermarks) support

Testing
./chip-tool softwarediagnostics reset-watermarks 1234 0

### CODE REVIEW

File: src/platform/CYW30739/DiagnosticDataProviderImpl.cpp [hunk 1]
```diff
@@ -124,6 +124,16 @@ void DiagnosticDataProviderImpl::ReleaseNetworkInterfaces(NetworkInterface * net
         Platform::Delete(del);
     }
 }
+
+CHIP_ERROR DiagnosticDataProviderImpl::ResetWatermarks()
+{
+    // If implemented, the server SHALL set the value of the CurrentHeapHighWatermark attribute to the
+    // value of the CurrentHeapUsed.
+    // On CYW30739, overide with non-op to pass CI.
```
Comment (bzbarsky-apple) [inline src/platform/CYW30739/DiagnosticDataProviderImpl.cpp:132]: This does not make sense.  CI should not be assuming this is supported, if this provider says it's not supported: this is an optional feature.  What CI is this trying to pass?

File: src/platform/CYW30739/DiagnosticDataProviderImpl.h [hunk 2]
```diff
@@ -39,11 +39,13 @@ class DiagnosticDataProviderImpl : public DiagnosticDataProvider
 
     // ===== Methods that implement the PlatformManager abstract interface.
 
+    bool SupportsWatermarks() override { return true; }
```
Comment (bzbarsky-apple) [inline src/platform/CYW30739/DiagnosticDataProviderImpl.h:42]: This should be returning false, given that this feature is not actually implemented.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

**Review by andy31415 — CHANGES_REQUESTED** [review]
Optional feature not being implemented is ok. This change is not needed (and likely harmful by claiming an unimplemented feature works).

**Comment by woody-apple** [comment]
> Optional feature not being implemented is ok. This change is not needed (and likely harmful by claiming an unimplemented feature works).

These are perfectly allowed in the SDK, they're just not mandatory. Approving

**Review by woody-apple — APPROVED** [review]

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
@andy31415 is right.  An actual implementation of an optional feature would be fine in the SDK.  But this PR claims to implement the feature, but actually no-ops the call so the feature is not in fact implemented.  That is not an OK thing to do; it's introducing a spec violation.

**Comment by AlvinHsiao** [comment]
@bzbarsky-apple We just added DGSW.S.F00(Watermarks) support to pass #3 "./chip-tool softwarediagnostics read feature-map 1  0" of TC-DGSW-1.1, or it would read the value "0" not expected "1".

**Comment by bzbarsky-apple** [comment]
@AlvinHsiao That test only expects 1 if the relevant PICS is claiming that the device supports watermarks, no?

**Comment by AlvinHsiao** [comment]
@bzbarsky-apple  Yes. But it seems no option to pass that test if the device not supports watermarks.

**Review by andy31415 — CHANGES_REQUESTED** [review]
Optional feature should either be implemented or not return 'supported'.

This PR marks it 'supported' and at the same time does a NOOP implementation,

**Comment by andy31415** [comment]
> @bzbarsky-apple Yes. But it seems no option to pass that test if the device not supports watermarks.

The test is wrong. If the spec says "Optional" you are allowed and spec compliant to not implement it.

**Comment by bzbarsky-apple** [comment]
> But it seems no option to pass that test if the device not supports watermarks.

No option in what sense?  How _exactly_ are you running the test?

**Comment by AlvinHsiao** [comment]
The test plan expects that the FeatureMap attribute has a value 1, but we got value 0 if we don't add "bool SupportsWatermarks() override { return true; }"

**Comment by bzbarsky-apple** [comment]
> The test plan expects that the FeatureMap attribute has a value 1

If the test plan expects that, the test plan is broken and _that_ is what we should fix.  I filed https://github.com/CHIP-Specifications/chip-test-plans/issues/1571 on the test plan.

**Pull request closed without merging**